### ISSUE TITLE
[Feature request] Gradient Accumulation

### ISSUE THREAD

**Issue by maksym-del** [allenai/allennlp#3469]
There are two PRs for this opened back in summer (https://github.com/allenai/allennlp/pull/2721, https://github.com/allenai/allennlp/pull/3051) but no open issue. This feature is crucial to reproduce finetuning results for some recent pretrained transformers. Is it anywhere on the radar to incorporate this feature?

**Comment by nelson-liu** [comment]
Agreed that this seems pretty crucial in the modern age of large LMs (and associated OOMs). It looks @ibeltagy also might have a version of this in his fork that was used for scibert?

**Comment by HarshTrivedi** [comment]
[This](https://github.com/allenai/allennlp/pull/3051) is usable pretty much as is if you are okay with using `CallbackTrainer `.

**Comment by DeNeutoy** [comment]
Yes, this is something we should just add to the default trainer. The implementation of this we have actually used internally in research projects is here: https://github.com/matt-peters/allennlp/commit/4992ed63137f8c66b0af4fbe3619ad0ca4085ccc

I'm not a fan of #3051 because it requires adding lots of variables to the base `CallbackTrainer` class, which was explicitly not the point of making the callback trainer in the first place.

I haven't examined the difference between the above link and #2721 . Does anyone have an opinion on the difference?

**Comment by DeNeutoy** [comment]
Overall we will likely get this done by the end of the year - if someone is willing to add this to specifically:

- The default trainer
- With some tests
- With an implementation that is agnostic to the iterator used by the trainer (i.e gradients from batches which are not a fixed size can be accumulated and applied)

We will review this quickly and be active in helping it into the library (cc @nelson-liu  😉 )

**Comment by brendan-ai2** [comment]
Previous issue on this topic: https://github.com/allenai/allennlp/issues/2112

**Comment by schmmd** [comment]
@DeNeutoy @brendan-ai2, @dirkgr is going to add gradient accumulation to the default trainer.

Also, if an external contributor is currently working on this please speak up so we can coordinate.

**Comment by dirkgr** [comment]
Done in #3537.
